File: packages/sardis-cli/src/sardis_cli/commands/mandates.py
"""CLI commands for spending mandate management."""
from __future__ import annotations

import click
from rich.console import Console
from rich.panel import Panel
from rich.table import Table

console = Console()


@click.group()
def mandates():
    """Manage spending mandates — scoped payment authorization for AI agents."""
    pass


@mandates.command("list")
@click.option("--status", type=click.Choice(["active", "draft", "suspended", "revoked", "expired", "consumed"]), help="Filter by status")
@click.option("--agent-id", help="Filter by agent ID")
@click.pass_context
def list_mandates(ctx, status, agent_id):
    """List spending mandates."""
    from sardis_cli.api import api_get

    params = {}
    if status:
        params["status_filter"] = status
    if agent_id:
        params["agent_id"] = agent_id

    data = api_get(ctx, "/api/v2/spending-mandates", params=params)
    if not data:
        console.print("[dim]No mandates found.[/dim]")
        return

    table = Table(title="Spending Mandates", show_lines=True)
    table.add_column("ID", style="cyan", no_wrap=True)
    table.add_column("Status", style="bold")
    table.add_column("Purpose")
    table.add_column("Per-TX", justify="right")
    table.add_column("Spent", justify="right", style="yellow")
    table.add_column("Total", justify="right")
    table.add_column("Approval")

    status_colors = {
        "active": "green", "draft": "dim", "suspended": "yellow",
        "revoked": "red", "expired": "dim", "consumed": "blue",
    }

    for m in data:
        s = m.get("status", "?")
        color = status_colors.get(s, "white")
        table.add_row(
            m.get("id", "?"),
            f"[{color}]{s}[/{color}]",
            (m.get("purpose_scope") or "—")[:40],
            f"${m.get('amount_per_tx', '∞')}",
            f"${m.get('spent_total', '0')}",
            f"${m.get('amount_total', '∞')}",
            m.get("approval_mode", "auto"),
        )

    console.print(table)


@mandates.command("create")
@click.option("--purpose", "-p", required=True, help="Purpose description")
@click.option("--per-tx", type=float, required=True, help="Max amount per transaction")
@click.option("--daily", type=float, help="Max daily spending")
@click.option("--monthly", type=float, help="Max monthly spending")
@click.option("--total", type=float, help="Total budget")
@click.option("--merchants", help="Allowed merchants (comma-separated)")
@click.option("--approval-threshold", type=float, help="Amount above which human approval needed")
@click.option("--approval-mode", type=click.Choice(["auto", "threshold", "always_human"]), default="auto")
@click.pass_context
def create_mandate(ctx, purpose, per_tx, daily, monthly, total, merchants, approval_threshold, approval_mode):
    """Create a new spending mandate."""
    from sardis_cli.api import api_post

    body = {
        "purpose_scope": purpose,
        "amount_per_tx": per_tx,
        "amount_daily": daily,
        "amount_monthly": monthly,
        "amount_total": total,
        "approval_threshold": approval_threshold,
        "approval_mode": approval_mode,
    }

    if merchants:
        body["merchant_scope"] = {"allowed": [m.strip() for m in merchants.split(",")]}

    result = api_post(ctx, "/api/v2/spending-mandates", body)
    if result:
        console.print(Panel(
            f"[green]Mandate created:[/green] {result.get('id', '?')}\n"
            f"Purpose: {purpose}\n"
            f"Per-TX: ${per_tx}\n"
            f"Status: {result.get('status', '?')}",
            title="✅ Spending Mandate Created",
        ))
    else:
        console.print("[red]Failed to create mandate.[/red]")


@mandates.command("revoke")
@click.argument("mandate_id")
@click.option("--reason", "-r", default="Revoked via CLI", help="Reason for revocation")
@click.pass_context
def revoke_mandate(ctx, mandate_id, reason):
    """Permanently revoke a spending mandate."""
    from sardis_cli.api import api_post

    if not click.confirm(f"Permanently revoke mandate {mandate_id}? This cannot be undone."):
        console.print("[dim]Cancelled.[/dim]")
        return

    result = api_post(ctx, f"/api/v2/spending-mandates/{mandate_id}/revoke", {"reason": reason})
    if result:
        console.print(f"[red]🚫 Mandate {mandate_id} permanently revoked.[/red]")
        console.print(f"Reason: {reason}")
    else:
        console.print(f"[red]Failed to revoke mandate {mandate_id}.[/red]")


@mandates.command("suspend")
@click.argument("mandate_id")
@click.option("--reason", "-r", default="Suspended via CLI", help="Reason")
@click.pass_context
def suspend_mandate(ctx, mandate_id, reason):
    """Temporarily suspend a spending mandate."""
    from sardis_cli.api import api_post

    result = api_post(ctx, f"/api/v2/spending-mandates/{mandate_id}/suspend", {"reason": reason})
    if result:
        console.print(f"[yellow]⏸️  Mandate {mandate_id} suspended.[/yellow]")
    else:
        console.print(f"[red]Failed to suspend mandate {mandate_id}.[/red]")


@mandates.command("resume")
@click.argument("mandate_id")
@click.pass_context
def resume_mandate(ctx, mandate_id):
    """Resume a suspended spending mandate."""
    from sardis_cli.api import api_post

    result = api_post(ctx, f"/api/v2/spending-mandates/{mandate_id}/resume", {})
    if result:
        console.print(f"[green]▶️  Mandate {mandate_id} resumed.[/green]")
    else:
        console.print(f"[red]Failed to resume mandate {mandate_id}.[/red]")

File: packages/sardis-mcp-server/src/tools/mandates.ts
/**
 * Spending Mandate tools for MCP server
 *
 * Tools for creating, managing, and validating spending mandates —
 * the core authorization primitive for AI agent payments.
 *
 * API base: /api/v2/spending-mandates (registered in sardis-api/main.py)
 * Query params: status_filter, agent_id (match API exactly)
 */

import { z } from 'zod';
import { getConfig } from '../config.js';
import { apiRequest } from '../api.js';
import type { ToolDefinition, ToolHandler, ToolResult } from './types.js';

// Tool definitions
export const mandateToolDefinitions: ToolDefinition[] = [
  {
    name: 'sardis_create_mandate',
    description:
      'Create a spending mandate — a scoped, revocable payment authorization for an AI agent. ' +
      'Defines what the agent can spend, on what merchants, up to what amount, with what approval rules.',
    inputSchema: {
      type: 'object',
      properties: {
        purpose: { type: 'string', description: 'Natural language description of what this mandate allows' },
        amount_per_tx: { type: 'number', description: 'Maximum amount per transaction' },
        amount_daily: { type: 'number', description: 'Maximum daily spending aggregate' },
        amount_monthly: { type: 'number', description: 'Maximum monthly spending aggregate' },
        amount_total: { type: 'number', description: 'Total lifetime budget for this mandate' },
        allowed_merchants: {
          type: 'array',
          items: { type: 'string' },
          description: 'List of allowed merchant domains (e.g., ["openai.com", "anthropic.com"])',
        },
        approval_threshold: { type: 'number', description: 'Amount above which human approval is required' },
        approval_mode: {
          type: 'string',
          enum: ['auto', 'threshold', 'always_human'],
          description: 'How approval works: auto (no approval), threshold (above amount), always_human (every payment)',
        },
        allowed_rails: {
          type: 'array',
          items: { type: 'string', enum: ['card', 'usdc', 'bank'] },
          description: 'Permitted payment rails',
        },
        agent_id: { type: 'string', description: 'Agent ID to bind this mandate to' },
      },
      required: ['purpose', 'amount_per_tx'],
    },
  },
  {
    name: 'sardis_list_mandates',
    description: 'List spending mandates for the organization. Filter by status or agent.',
    inputSchema: {
      type: 'object',
      properties: {
        status: {
          type: 'string',
          enum: ['active', 'draft', 'suspended', 'revoked', 'expired', 'consumed'],
          description: 'Filter by mandate status',
        },
        agent_id: { type: 'string', description: 'Filter by agent ID' },
      },
      required: [],
    },
  },
  {
    name: 'sardis_revoke_mandate',
    description:
      'Permanently revoke a spending mandate. This is irreversible — the mandate cannot be reactivated. ' +
      'All future payments under this mandate will be blocked immediately.',
    inputSchema: {
      type: 'object',
      properties: {
        mandate_id: { type: 'string', description: 'ID of the mandate to revoke' },
        reason: { type: 'string', description: 'Reason for revoking the mandate' },
      },
      required: ['mandate_id'],
    },
  },
  {
    name: 'sardis_check_mandate',
    description:
      'Dry-run: check if a payment would be authorized by a specific spending mandate. ' +
      'Returns whether the payment would be approved, rejected, or require human approval.',
    inputSchema: {
      type: 'object',
      properties: {
        mandate_id: { type: 'string', description: 'ID of the mandate to check against' },
        amount: { type: 'number', description: 'Payment amount to check' },
        merchant: { type: 'string', description: 'Merchant domain to check' },
        rail: {
          type: 'string',
          enum: ['card', 'usdc', 'bank'],
          description: 'Payment rail to check',
        },
      },
      required: ['mandate_id', 'amount'],
    },
  },
];

// Tool handlers
async function handleCreateMandate(args: unknown): Promise<ToolResult> {
  const input = args as Record<string, unknown>;
  const config = getConfig();

  const body: Record<string, unknown> = {
    purpose_scope: input.purpose,
    amount_per_tx: input.amount_per_tx,
    amount_daily: input.amount_daily,
    amount_monthly: input.amount_monthly,
    amount_total: input.amount_total,
    approval_threshold: input.approval_threshold,
    approval_mode: input.approval_mode || 'auto',
    allowed_rails: input.allowed_rails || ['card', 'usdc', 'bank'],
    agent_id: input.agent_id,
  };

  if (input.allowed_merchants) {
    body.merchant_scope = { allowed: input.allowed_merchants };
  }

  if (!config.apiKey) {
    // Simulated mode
    const id = `mandate_sim_${Date.now().toString(36)}`;
    return {
      content: [
        {
          type: 'text',
          text: JSON.stringify({
            _simulated: true,
            _warning: 'This is simulated data. Configure SARDIS_API_KEY for real data.',
            id,
            purpose: input.purpose,
            amount_per_tx: input.amount_per_tx,
            status: 'active',
            message: `Spending mandate created (simulated). Connect an API key for real mandates.`,
          }, null, 2),
        },
      ],
    };
  }

  const result = await apiRequest<Record<string, unknown>>('POST', '/api/v2/spending-mandates', body);
  return {
    content: [
      {
        type: 'text',
        text: `✅ Spending mandate created\n\nID: ${result.id}\nPurpose: ${result.purpose_scope}\nPer-tx limit: $${result.amount_per_tx}\nApproval mode: ${result.approval_mode}\nStatus: ${result.status}`,
      },
    ],
  };
}

async function handleListMandates(args: unknown): Promise<ToolResult> {
  const input = args as Record<string, unknown>;
  const config = getConfig();

  if (!config.apiKey) {
    return {
      content: [
        {
          type: 'text',
          text: JSON.stringify({
            _simulated: true,
            _warning: 'This is simulated data. Configure SARDIS_API_KEY for real data.',
            mandates: [],
            message: 'No mandates (simulated mode). Connect an API key to manage real mandates.',
          }, null, 2),
        },
      ],
    };
  }

  let url = '/api/v2/spending-mandates';
  const params: string[] = [];
  if (input.status) params.push(`status_filter=${input.status}`);
  if (input.agent_id) params.push(`agent_id=${input.agent_id}`);
  if (params.length) url += `?${params.join('&')}`;

  const mandates = await apiRequest<Record<string, unknown>[]>('GET', url);

  if (!mandates.length) {
    return { content: [{ type: 'text', text: '📋 No spending mandates found.' }] };
  }

  const lines = mandates.map(
    (m) =>
      `• ${m.id} [${m.status}] — ${m.purpose_scope || 'No purpose'} | Per-tx: $${m.amount_per_tx || '∞'} | Spent: $${m.spent_total || '0'}`
  );

  return {
    content: [{ type: 'text', text: `📋 Spending Mandates (${mandates.length})\n\n${lines.join('\n')}` }],
  };
}

async function handleRevokeMandate(args: unknown): Promise<ToolResult> {
  const input = args as Record<string, unknown>;
  const config = getConfig();

  if (!config.apiKey) {
    return {
      content: [{
        type: 'text',
        text: JSON.stringify({
          _simulated: true,
          _warning: 'This is simulated data. Configure SARDIS_API_KEY for real data.',
          mandate_id: input.mandate_id,
          status: 'revoked',
          message: `Mandate ${input.mandate_id} revoked (simulated).`,
        }, null, 2),
      }],
    };
  }

  await apiRequest<Record<string, unknown>>('POST', `/api/v2/spending-mandates/${input.mandate_id}/revoke`, {
    reason: input.reason || 'Revoked via MCP',
  });

  return {
    content: [
      {
        type: 'text',
        text: `🚫 Mandate ${input.mandate_id} permanently revoked.\nReason: ${input.reason || 'Revoked via MCP'}\n\nAll future payments under this mandate are now blocked.`,
      },
    ],
  };
}

async function handleCheckMandate(args: unknown): Promise<ToolResult> {
  const input = args as Record<string, unknown>;
  const config = getConfig();

  if (!config.apiKey) {
    return {
      content: [
        {
          type: 'text',
          text: JSON.stringify({
            _simulated: true,
            _warning: 'This is simulated data. Configure SARDIS_API_KEY for real data.',
            amount: input.amount,
            authorized: true,
            message: `Payment of $${input.amount} would be authorized (simulated). Connect an API key for real mandate validation.`,
          }, null, 2),
        },
      ],
    };
  }

  // Get the mandate details to check locally
  const mandate = await apiRequest<Record<string, unknown>>('GET', `/api/v2/spending-mandates/${input.mandate_id}`);

  const perTx = mandate.amount_per_tx ? parseFloat(mandate.amount_per_tx as string) : Infinity;
  const amount = input.amount as number;
  const status = mandate.status as string;

  if (status !== 'active') {
    return {
      content: [{ type: 'text', text: `❌ Mandate is ${status} — payment would be rejected.` }],
    };
  }

  if (amount > perTx) {
    return {
      content: [
        {
          type: 'text',
          text: `❌ Payment of $${amount} exceeds per-transaction limit of $${perTx}.\n\nSuggestion: Reduce amount or update mandate limits.`,
        },
      ],
    };
  }

  const threshold = mandate.approval_threshold ? parseFloat(mandate.approval_threshold as string) : Infinity;
  const needsApproval = amount > threshold;

  return {
    content: [
      {
        type: 'text',
        text: `✅ Payment of $${amount} would be authorized.${needsApproval ? '\n⚠️ Human approval required (above threshold).' : ''}`,
      },
    ],
  };
}

export const mandateToolHandlers: Record<string, ToolHandler> = {
  sardis_create_mandate: handleCreateMandate,
  sardis_list_mandates: handleListMandates,
  sardis_revoke_mandate: handleRevokeMandate,
  sardis_check_mandate: handleCheckMandate,
};
